.github/workflows/ci.yaml

```diff
@@ -18,44 +18,131 @@ jobs:
       run: echo ${{ github.ref }}
     - name: event_name
       run: echo ${{ github.event_name }}
-  build:
-    name: Build
+  lint:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v3
+    - uses: actions/setup-go@v2
+      with:
+        go-version: 1.17
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v3.1.0
+      with:
+        version: v1.44.2
+  test:
+    name: Test
     runs-on: ubuntu-latest
     steps:
     - name: checkout
-      uses: actions/checkout@v2
+      uses: actions/checkout@v3
+
     - uses: actions/setup-go@v2
       with:
-        go-version: '1.17' # The Go version to download (if necessary) and use.
-    - name: ci
-      run: make ci
+        go-version: 1.17
+    - id: go-cache-paths
+      shell: bash
+      run: |
+        echo "::set-output name=go-build::$(go env GOCACHE)"
+        echo "::set-output name=go-mod::$(go env GOMODCACHE)"
+    - name: Go Mod Cache
+      uses: actions/cache@v2
+      with:
+        path: ${{ steps.go-cache-paths.outputs.go-mod }}
+        key: ${{ runner.os }}-go-mod-${{ hashFiles('**/go.sum') }}
+        restore-keys: |
+          ${{ runner.os }}-go-mod-
+    - name: Go Build Cache
+      uses: actions/cache@v2
+      with:
+        path: ${{ steps.go-cache-paths.outputs.go-build }}
+        key: ${{ runner.os }}-go-build-${{ hashFiles('**/go.sum') }}
+        restore-keys: |
+          ${{ runner.os }}-go-build-
+
+    - name: test
+      run: make race
+
+  image:
+    name: Build Image
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v3
     - name: Set up QEMU for cross-building
       uses: docker/setup-qemu-action@v1
     - name: Set up Docker Buildx
       uses: docker/setup-buildx-action@v1
-    - name: Build
-      id: docker_build
-      uses: docker/build-push-action@v2
+
+    - name: Build image urls
+      id: image-urls
+      env:
+        DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
+        QUAY_USERNAME: ${{ secrets.QUAY_USERNAME }}
+      run: |
+        IMAGES=ghcr.io/${IMAGE_NAME}
+        [[ -n "$DOCKER_USERNAME" ]] && IMAGES=${IMAGES},${IMAGE_NAME}
+        [[ -n "$QUAY_USERNAME" ]] && IMAGES=${IMAGES},quay.io/${IMAGE_NAME}
+        echo "::set-output name=images::${IMAGES}"
```

Comment:
By default use ghcr.io, since we know we have credentials locally for that. Only add docker and quay.io registries when there are secrets available for them.